diff --git a/crypto/obj/obj_dat.h b/crypto/obj/obj_dat.h
index feb8f2d1e..e7a00f46c 100644
--- a/crypto/obj/obj_dat.h
+++ b/crypto/obj/obj_dat.h
@@ -8800,7 +8800,8 @@ static const ASN1_OBJECT kObjects[NUM_NID] = {
     {"id-ml-dsa-87", "ML-DSA-87", NID_ML_DSA_87, 9, &kObjectData[6223], 0},
     {"id-alg-ml-kem-768", "ML-KEM-768", NID_ML_KEM_768, 9, &kObjectData[6232],
      0},
-    {NULL, NULL, NID_undef, 0, NULL, 0},
+    {"P256Kyber768Draft00", "P256Kyber768Draft00", NID_P256Kyber768Draft00, 0,
+     NULL, 0},
     {"X-Wing", "X-Wing", NID_X_Wing, 0, NULL, 0},
 };
 
@@ -8934,6 +8935,7 @@ static const uint16_t kNIDsInShortNameOrder[] = {
     18 /* OU */,
     749 /* Oakley-EC2N-3 */,
     750 /* Oakley-EC2N-4 */,
+    971 /* P256Kyber768Draft00 */,
     9 /* PBE-MD2-DES */,
     168 /* PBE-MD2-RC2-64 */,
     10 /* PBE-MD5-DES */,
@@ -9858,6 +9860,7 @@ static const uint16_t kNIDsInLongNameOrder[] = {
     366 /* OCSP Nonce */,
     371 /* OCSP Service Locator */,
     180 /* OCSP Signing */,
+    971 /* P256Kyber768Draft00 */,
     161 /* PBES2 */,
     69 /* PBKDF2 */,
     162 /* PBMAC1 */,
diff --git a/crypto/obj/obj_mac.num b/crypto/obj/obj_mac.num
index 7b8bd2ead..25eac5075 100644
--- a/crypto/obj/obj_mac.num
+++ b/crypto/obj/obj_mac.num
@@ -958,4 +958,5 @@ ML_DSA_44		967
 ML_DSA_65		968
 ML_DSA_87		969
 ML_KEM_768		970
+P256Kyber768Draft00		971
 X_Wing		972
diff --git a/crypto/obj/objects.txt b/crypto/obj/objects.txt
index 67b9ade43..384b25753 100644
--- a/crypto/obj/objects.txt
+++ b/crypto/obj/objects.txt
@@ -1340,6 +1340,7 @@ secg-scheme 14 3 : dhSinglePass-cofactorDH-sha512kdf-scheme
 
 # NIDs for post quantum hybrid KEMs in TLS (no corresponding OIDs).
  : X25519Kyber768Draft00
+ : P256Kyber768Draft00
  : X25519MLKEM768
 
 # NIDs for PQ/T hybrid KEMs (no corresponding OIDs).
diff --git a/include/openssl/nid.h b/include/openssl/nid.h
index 6b3bb4506..09af03d11 100644
--- a/include/openssl/nid.h
+++ b/include/openssl/nid.h
@@ -5508,6 +5508,9 @@ extern "C" {
 #define OBJ_ML_KEM_768 2L, 16L, 840L, 1L, 101L, 3L, 4L, 4L, 2L
 #define OBJ_ENC_ML_KEM_768 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x04, 0x02
 
+#define SN_P256Kyber768Draft00 "P256Kyber768Draft00"
+#define NID_P256Kyber768Draft00 971
+
 #define SN_X_Wing "X-Wing"
 #define NID_X_Wing 972
 
diff --git a/include/openssl/ssl.h b/include/openssl/ssl.h
index 56682da6a..8a7c717eb 100644
--- a/include/openssl/ssl.h
+++ b/include/openssl/ssl.h
@@ -2620,6 +2620,7 @@ OPENSSL_EXPORT size_t SSL_CTX_get_num_tickets(const SSL_CTX *ctx);
 #define SSL_GROUP_X25519_MLKEM768 0x11ec
 #define SSL_GROUP_X25519_KYBER768_DRAFT00 0x6399
 #define SSL_GROUP_MLKEM1024 0x0202
+#define SSL_GROUP_P256_KYBER768_DRAFT00 0xfe32
 
 // SSL_CTX_set1_group_ids sets the preferred groups for `ctx` to `group_ids`.
 // Each element of `group_ids` should be a unique one of the `SSL_GROUP_*`
@@ -6286,6 +6287,20 @@ OPENSSL_EXPORT int SSL_CTX_set1_curves_list(SSL_CTX *ctx, const char *curves);
 // SSL_set1_curves_list calls `SSL_set1_groups_list`.
 OPENSSL_EXPORT int SSL_set1_curves_list(SSL *ssl, const char *curves);
 
+// By default, a client will send both a non post-quantum and a post-quantum
+// keyshare if available.
+//
+// SSL_use_second_keyshare controls this behaviour. If |enabled| is 0, then
+// a client using |ssl| will only send one keyshare.
+OPENSSL_EXPORT void SSL_use_second_keyshare(SSL *ssl, int enabled);
+
+// By default, a client will send both a non post-quantum and a post-quantum
+// keyshare if available.
+//
+// SSL_CTX_use_second_keyshare controls this behaviour. If |enabled| is 0, then
+// a client using |ctx| will only send one keyshare.
+OPENSSL_EXPORT void SSL_CTX_use_second_keyshare(SSL_CTX *ctx, int enabled);
+
 // TLSEXT_nid_unknown is a constant used in OpenSSL for
 // `SSL_get_negotiated_group` to return an unrecognized group. BoringSSL never
 // returns this value, but we define this constant for compatibility.
diff --git a/ssl/extensions.cc b/ssl/extensions.cc
index 8787d9fe4..56de76e37 100644
--- a/ssl/extensions.cc
+++ b/ssl/extensions.cc
@@ -109,6 +109,7 @@ static bool tls1_check_duplicate_extensions(const CBS *cbs) {
 static bool is_post_quantum_group(uint16_t id) {
   switch (id) {
     case SSL_GROUP_X25519_KYBER768_DRAFT00:
+    case SSL_GROUP_P256_KYBER768_DRAFT00:
     case SSL_GROUP_X25519_MLKEM768:
     case SSL_GROUP_MLKEM1024:
       return true;
@@ -2471,18 +2472,21 @@ bool ssl_setup_key_shares(SSL_HANDSHAKE *hs, uint16_t override_group_id) {
     if (!default_key_shares.TryPushBack(supported_group_list[0])) {
       return false;
     }
-    // We'll try to include one post-quantum and one classical initial key
-    // share.
-    for (size_t i = 1; i < supported_group_list.size(); i++) {
-      if (is_post_quantum_group(default_key_shares[0]) ==
-          is_post_quantum_group(supported_group_list[i])) {
-        continue;
-      }
-      if (!default_key_shares.TryPushBack(supported_group_list[i])) {
-        return false;
+
+    if (!ssl->config->disable_second_keyshare) {
+      // We'll try to include one post-quantum and one classical initial key
+      // share.
+      for (size_t i = 1; i < supported_group_list.size(); i++) {
+        if (is_post_quantum_group(default_key_shares[0]) ==
+            is_post_quantum_group(supported_group_list[i])) {
+          continue;
+        }
+        if (!default_key_shares.TryPushBack(supported_group_list[i])) {
+          return false;
+        }
+        assert(default_key_shares[1] != default_key_shares[0]);
+        break;
       }
-      assert(default_key_shares[1] != default_key_shares[0]);
-      break;
     }
     selected_key_shares.emplace(default_key_shares);
   }
diff --git a/ssl/internal.h b/ssl/internal.h
index 74c6c2e9c..a8643452f 100644
--- a/ssl/internal.h
+++ b/ssl/internal.h
@@ -918,7 +918,7 @@ struct NamedGroup {
 Span<const NamedGroup> NamedGroups();
 
 // kNumNamedGroups is the number of supported groups.
-constexpr size_t kNumNamedGroups = 7u;
+constexpr size_t kNumNamedGroups = 8u;
 
 // DefaultSupportedGroupIds returns the list of IDs for the default groups that
 // are supported when the caller hasn't explicitly configured supported groups.
@@ -3541,6 +3541,11 @@ struct SSL_CONFIG {
   // permute_extensions is whether to permute extensions when sending messages.
   bool permute_extensions : 1;
 
+  // As a client by default we will send a non post-quantum share and
+  // a post-quantum share if available. If disable_second_keyshare is set,
+  // we will only send the most preferred keyshare.
+  bool disable_second_keyshare : 1;
+
   // aes_hw_override if set indicates we should override checking for aes
   // hardware support, and use the value in aes_hw_override_value instead.
   bool aes_hw_override : 1;
@@ -4192,6 +4197,11 @@ class SSLContext : public ssl_ctx_st, public RefCounted<SSLContext> {
   // permute_extensions is whether to permute extensions when sending messages.
   bool permute_extensions : 1;
 
+  // As a client by default we will send a non post-quantum share and
+  // a post-quantum share if available. If disable_second_keyshare is set,
+  // we will only send the most preferred keyshare.
+  bool disable_second_keyshare : 1;
+
   // allow_unknown_alpn_protos is whether the client allows unsolicited ALPN
   // protocols from the peer.
   bool allow_unknown_alpn_protos : 1;
diff --git a/ssl/ssl_key_share.cc b/ssl/ssl_key_share.cc
index 841a6df0d..a6fec6ae8 100644
--- a/ssl/ssl_key_share.cc
+++ b/ssl/ssl_key_share.cc
@@ -193,6 +193,109 @@ class X25519KeyShare : public SSLKeyShare {
   uint8_t private_key_[32];
 };
 
+class P256Kyber768Draft00KeyShare : public SSLKeyShare {
+ public:
+  P256Kyber768Draft00KeyShare()
+    : ecks_(EC_group_p256(), SSL_GROUP_SECP256R1) {}
+
+  uint16_t GroupID() const override {
+    return SSL_GROUP_P256_KYBER768_DRAFT00;
+  }
+
+  bool Generate(CBB *out) override {
+    uint8_t kyber_public_key[KYBER_PUBLIC_KEY_BYTES];
+    KYBER_generate_key(kyber_public_key, &kyber_private_key_);
+
+    if(!ecks_.Generate(out) ||
+       !CBB_add_bytes(out, kyber_public_key, sizeof(kyber_public_key))) {
+      return false;
+    }
+
+    return true;
+  }
+
+  bool Encap(CBB *out_ciphertext, Array<uint8_t> *out_secret,
+             uint8_t *out_alert, Span<const uint8_t> peer_key) override {
+    Array<uint8_t> ec_secret;
+
+    *out_alert = SSL_AD_INTERNAL_ERROR;
+
+    if(peer_key.size() != p256_share_size + KYBER_PUBLIC_KEY_BYTES) {
+      *out_alert = SSL_AD_ILLEGAL_PARAMETER;
+      OPENSSL_PUT_ERROR(SSL, SSL_R_BAD_ECPOINT);
+      return false;
+    }
+
+    if (!ecks_.Encap(out_ciphertext, &ec_secret, out_alert,
+                    peer_key.subspan(0, p256_share_size))) {
+      return false;
+    }
+
+    KYBER_public_key peer_kyber_pub;
+    CBS peer_kyber_cbs;
+    CBS_init(&peer_kyber_cbs, peer_key.data() + p256_share_size,
+             KYBER_PUBLIC_KEY_BYTES);
+
+    if (!KYBER_parse_public_key(&peer_kyber_pub, &peer_kyber_cbs)) {
+      *out_alert = SSL_AD_ILLEGAL_PARAMETER;
+      OPENSSL_PUT_ERROR(SSL, SSL_R_BAD_ECPOINT);
+      return false;
+    }
+
+    uint8_t kyber_ciphertext[KYBER_CIPHERTEXT_BYTES];
+    Array<uint8_t> secret;
+    if (!secret.InitForOverwrite(p256_secret_size + KYBER_SHARED_SECRET_BYTES)) {
+      return false;
+    }
+    OPENSSL_memcpy(secret.data(), ec_secret.data(), ec_secret.size());
+    KYBER_encap(kyber_ciphertext, secret.data() + p256_secret_size,
+                &peer_kyber_pub);
+
+    if(!CBB_add_bytes(out_ciphertext, kyber_ciphertext,
+                       sizeof(kyber_ciphertext))) {
+      return false;
+    }
+
+    *out_secret = std::move(secret);
+    return true;
+  }
+
+  bool Decap(Array<uint8_t> *out_secret, uint8_t *out_alert,
+             Span<const uint8_t> ciphertext) override {
+    *out_alert = SSL_AD_INTERNAL_ERROR;
+
+    Array<uint8_t> ec_secret;
+
+    if (ciphertext.size() != p256_share_size + KYBER_CIPHERTEXT_BYTES) {
+      *out_alert = SSL_AD_ILLEGAL_PARAMETER;
+      OPENSSL_PUT_ERROR(SSL, SSL_R_BAD_ECPOINT);
+      return false;
+    }
+
+    if (!ecks_.Decap(&ec_secret, out_alert,
+                     ciphertext.subspan(0, p256_share_size))) {
+      return false;
+    }
+
+    Array<uint8_t> secret;
+    if (!secret.InitForOverwrite(p256_secret_size + KYBER_SHARED_SECRET_BYTES)) {
+      return false;
+    }
+    OPENSSL_memcpy(secret.data(), ec_secret.data(), ec_secret.size());
+    KYBER_decap(secret.data() + p256_secret_size,
+                ciphertext.data() + p256_share_size, &kyber_private_key_);
+    *out_secret = std::move(secret);
+    return true;
+  }
+
+ private:
+  ECKeyShare ecks_;
+  KYBER_private_key kyber_private_key_;
+
+  static constexpr size_t p256_share_size = 65;
+  static constexpr size_t p256_secret_size = 32;
+};
+
 // draft-tls-westerbaan-xyber768d00-03
 class X25519Kyber768KeyShare : public SSLKeyShare {
  public:
@@ -441,9 +544,11 @@ constexpr NamedGroup kNamedGroups[] = {
     {NID_secp521r1, SSL_GROUP_SECP521R1, "P-521", "secp521r1"},
     {NID_X25519, SSL_GROUP_X25519, "X25519", "x25519"},
     {NID_X25519Kyber768Draft00, SSL_GROUP_X25519_KYBER768_DRAFT00,
-     "X25519Kyber768Draft00", ""},
+     "X25519Kyber768Draft00", "Xyber768D00"},
     {NID_X25519MLKEM768, SSL_GROUP_X25519_MLKEM768, "X25519MLKEM768", ""},
     {NID_ML_KEM_1024, SSL_GROUP_MLKEM1024, "MLKEM1024", ""},
+    {NID_P256Kyber768Draft00, SSL_GROUP_P256_KYBER768_DRAFT00,
+     "P256Kyber768Draft00", "P256Kyber768D00"},
 };
 
 static_assert(std::size(kNamedGroups) == kNumNamedGroups,
@@ -456,6 +561,7 @@ Span<const NamedGroup> NamedGroups() { return kNamedGroups; }
 Span<const uint16_t> DefaultSupportedGroupIds() {
   static const uint16_t kDefaultSupportedGroupIds[] = {
       SSL_GROUP_X25519_MLKEM768,
+      SSL_GROUP_P256_KYBER768_DRAFT00,
       SSL_GROUP_X25519,
       SSL_GROUP_SECP256R1,
       SSL_GROUP_SECP384R1,
@@ -479,6 +585,8 @@ UniquePtr<SSLKeyShare> SSLKeyShare::Create(uint16_t group_id) {
       return MakeUnique<X25519MLKEM768KeyShare>();
     case SSL_GROUP_MLKEM1024:
       return MakeUnique<MLKEM1024KeyShare>();
+    case SSL_GROUP_P256_KYBER768_DRAFT00:
+      return MakeUnique<P256Kyber768Draft00KeyShare>();
     default:
       return nullptr;
   }
diff --git a/ssl/ssl_lib.cc b/ssl/ssl_lib.cc
index e05f91bd9..07e6ff069 100644
--- a/ssl/ssl_lib.cc
+++ b/ssl/ssl_lib.cc
@@ -394,6 +394,7 @@ bssl::SSLContext::SSLContext(const SSL_METHOD *ssl_method)
       grease_enabled(false),
       grease_sigalgs_enabled(false),
       permute_extensions(false),
+      disable_second_keyshare(false),
       allow_unknown_alpn_protos(false),
       false_start_allowed_without_alpn(false),
       handoff(false),
@@ -527,6 +528,7 @@ SSL *SSL_new(SSL_CTX *ctx) {
   ssl->config->retain_only_sha256_of_client_certs =
       ctx_impl->retain_only_sha256_of_client_certs;
   ssl->config->permute_extensions = ctx_impl->permute_extensions;
+  ssl->config->disable_second_keyshare = ctx_impl->disable_second_keyshare;
   ssl->config->aes_hw_override = ctx_impl->aes_hw_override;
   ssl->config->aes_hw_override_value = ctx_impl->aes_hw_override_value;
   ssl->config->compliance_policy = ctx_impl->compliance_policy;
@@ -592,6 +594,7 @@ SSL_CONFIG::SSL_CONFIG(SSLImpl *ssl_arg)
       jdk11_workaround(false),
       quic_use_legacy_codepoint(false),
       permute_extensions(false),
+      disable_second_keyshare(false),
       alps_use_new_codepoint(true),
       server_padding_enabled(false) {
   assert(ssl);
@@ -3526,6 +3529,19 @@ int SSL_set1_curves_list(SSL *ssl, const char *curves) {
   return SSL_set1_groups_list(ssl, curves);
 }
 
+void SSL_use_second_keyshare(SSL *ssl, int enabled) {
+  auto *ssl_impl = FromOpaque(ssl);
+  if (!ssl_impl->config) {
+    return;
+  }
+  ssl_impl->config->disable_second_keyshare = !enabled;
+}
+
+void SSL_CTX_use_second_keyshare(SSL_CTX *ctx, int enabled) {
+  FromOpaque(ctx)->disable_second_keyshare = !enabled;
+}
+
+
 namespace fips202205 {
 
 // (References are to SP 800-52r2):
@@ -3537,7 +3549,8 @@ namespace fips202205 {
 // Section 3.3.1
 // "The server shall be configured to only use cipher suites that are
 // composed entirely of NIST approved algorithms"
-static const uint16_t kGroups[] = {SSL_GROUP_SECP256R1, SSL_GROUP_SECP384R1};
+static const uint16_t kGroups[] = {SSL_GROUP_P256_KYBER768_DRAFT00,
+                                   SSL_GROUP_SECP256R1, SSL_GROUP_SECP384R1};
 
 static const uint16_t kSigAlgs[] = {
     SSL_SIGN_RSA_PKCS1_SHA256,
diff --git a/ssl/ssl_test.cc b/ssl/ssl_test.cc
index b1a8941c1..a8e308398 100644
--- a/ssl/ssl_test.cc
+++ b/ssl/ssl_test.cc
@@ -524,6 +524,14 @@ static const CurveTest kCurveTests[] = {
         "MLKEM1024:X25519MLKEM768",
         {SSL_GROUP_MLKEM1024, SSL_GROUP_X25519_MLKEM768},
     },
+    {
+        "P256Kyber768Draft00",
+        {SSL_GROUP_P256_KYBER768_DRAFT00},
+    },
+    {
+        "P-256:P256Kyber768Draft00",
+        {SSL_GROUP_SECP256R1, SSL_GROUP_P256_KYBER768_DRAFT00},
+    },
 
     {
         "P-256:P-384:P-521:X25519",
@@ -686,8 +694,9 @@ TEST(SSLTest, CurveRules) {
 }
 
 TEST(SSLTest, DefaultCurves) {
-  const uint16_t kDefaults[] = {SSL_GROUP_X25519_MLKEM768, SSL_GROUP_X25519,
-                                SSL_GROUP_SECP256R1, SSL_GROUP_SECP384R1};
+  const uint16_t kDefaults[] = {
+      SSL_GROUP_X25519_MLKEM768, SSL_GROUP_P256_KYBER768_DRAFT00,
+      SSL_GROUP_X25519, SSL_GROUP_SECP256R1, SSL_GROUP_SECP384R1};
 
   // Test the group ID APIs.
   {
@@ -7192,7 +7201,9 @@ TEST(SSLTest, ApplyHandoffRemovesUnsupportedCurves) {
 
   // The default list of groups is used before applying the handoff.
   EXPECT_THAT(FromOpaque(server.get())->config->supported_group_list,
-              ElementsAreArray({SSL_GROUP_X25519_MLKEM768, SSL_GROUP_X25519,
+              ElementsAreArray({SSL_GROUP_X25519_MLKEM768,
+                                SSL_GROUP_P256_KYBER768_DRAFT00,
+                                SSL_GROUP_X25519,
                                 SSL_GROUP_SECP256R1, SSL_GROUP_SECP384R1}));
   ASSERT_TRUE(SSL_apply_handoff(server.get(), handoff));
   EXPECT_EQ(1u, FromOpaque(server.get())->config->supported_group_list.size());
diff --git a/ssl/test/runner/cbc_tests.go b/ssl/test/runner/cbc_tests.go
index 6f49d12af..c17e3f277 100644
--- a/ssl/test/runner/cbc_tests.go
+++ b/ssl/test/runner/cbc_tests.go
@@ -14,6 +14,8 @@
 
 package runner
 
+import "strconv"
+
 func addCBCPaddingTests() {
 	testCases = append(testCases, testCase{
 		name: "MaxCBCPadding",
@@ -104,6 +106,7 @@ func addCBCSplittingTests() {
 				"-partial-write",
 				// BoringSSL disables 3DES by default.
 				"-cipher", "ALL:3DES",
+				"-curves", strconv.Itoa(int(CurveX25519)),
 			},
 		})
 	}
diff --git a/ssl/test/runner/key_update_tests.go b/ssl/test/runner/key_update_tests.go
index 39ad24baf..f7335bb9a 100644
--- a/ssl/test/runner/key_update_tests.go
+++ b/ssl/test/runner/key_update_tests.go
@@ -16,6 +16,7 @@ package runner
 
 import (
-	"slices"
+	"slices"
+	"strconv"
 )
 
 func addKeyUpdateTests() {
@@ -296,6 +297,7 @@ func addKeyUpdateTests() {
 			},
 		},
 		shimSendsKeyUpdateBeforeRead: true,
+		flags:                        []string{"-curves", strconv.Itoa(int(CurveX25519))},
 	})
 
 	// Test that shim responds to KeyUpdate requests.
diff --git a/tool/client.cc b/tool/client.cc
index 1f2b19965..a69d90aca 100644
--- a/tool/client.cc
+++ b/tool/client.cc
@@ -156,6 +156,11 @@ static const struct argument kArguments[] = {
         kBooleanArgument,
         "Permute extensions in handshake messages",
     },
+    {
+        "-disable-second-keyshare",
+        kBooleanArgument,
+        "Do not send a second keyshare",
+    },
     {
         "-test-resumption",
         kBooleanArgument,
@@ -241,8 +246,9 @@ static const struct argument kArguments[] = {
     },
 };
 
-static int NextProtoSelectCallback(SSL* ssl, uint8_t** out, uint8_t* outlen,
-                                   const uint8_t* in, unsigned inlen, void* arg) {
+static int NextProtoSelectCallback(SSL *ssl, uint8_t **out, uint8_t *outlen,
+                                   const uint8_t *in, unsigned inlen,
+                                   void *arg) {
   *out = reinterpret_cast<uint8_t *>(arg);
   *outlen = strlen(reinterpret_cast<const char *>(arg));
   return SSL_TLSEXT_ERR_OK;
@@ -325,7 +331,7 @@ static bool DoConnection(SSL_CTX *ctx,
   }
 
   if (args_map.count("-starttls") != 0) {
-    const std::string& starttls = args_map["-starttls"];
+    const std::string &starttls = args_map["-starttls"];
     if (starttls == "smtp") {
       if (!DoSMTPStartTLS(sock)) {
         return false;
@@ -362,15 +368,15 @@ static bool DoConnection(SSL_CTX *ctx,
   }
 
   if (args_map.count("-session-in") != 0) {
-    bssl::UniquePtr<BIO> in(BIO_new_file(args_map["-session-in"].c_str(),
-                                         "rb"));
+    bssl::UniquePtr<BIO> in(
+        BIO_new_file(args_map["-session-in"].c_str(), "rb"));
     if (!in) {
       fprintf(stderr, "Error reading session\n");
       ERR_print_errors_fp(stderr);
       return false;
     }
-    bssl::UniquePtr<SSL_SESSION> session(PEM_read_bio_SSL_SESSION(in.get(),
-                                         nullptr, nullptr, nullptr));
+    bssl::UniquePtr<SSL_SESSION> session(
+        PEM_read_bio_SSL_SESSION(in.get(), nullptr, nullptr, nullptr));
     if (!session) {
       fprintf(stderr, "Error reading session\n");
       ERR_print_errors_fp(stderr);
@@ -637,6 +643,10 @@ bool Client(const std::vector<std::string> &args) {
     SSL_CTX_set_permute_extensions(ctx.get(), 1);
   }
 
+  if (args_map.count("-disable-second-keyshare") != 0) {
+    SSL_CTX_use_second_keyshare(ctx.get(), 0);
+  }
+
   // Configure accepted roots.
   if (args_map.count("-root-certs") != 0) {
     if (!SSL_CTX_load_verify_locations(
@@ -648,8 +658,8 @@ bool Client(const std::vector<std::string> &args) {
     SSL_CTX_set_verify(ctx.get(), SSL_VERIFY_PEER, nullptr);
   }
   if (args_map.count("-root-cert-dir") != 0) {
-    if (!SSL_CTX_load_verify_locations(
-            ctx.get(), nullptr, args_map["-root-cert-dir"].c_str())) {
+    if (!SSL_CTX_load_verify_locations(ctx.get(), nullptr,
+                                       args_map["-root-cert-dir"].c_str())) {
       fprintf(stderr, "Failed to load root certificates.\n");
       ERR_print_errors_fp(stderr);
       return false;
diff --git a/ssl/ssl_test.cc b/ssl/ssl_test.cc
index 88b81e35a..4d0f3252c 100644
--- a/ssl/ssl_test.cc
+++ b/ssl/ssl_test.cc
@@ -2359,9 +2374,9 @@ TEST(SSLTest, SetGroupIdsWithEqualPreference) {
 // Test that the SSL group flags are defaulted to zero when zero groups are set
 // (i.e. using the default groups).
 TEST(SSLTest, SetGroupIdsWithFlags_DefaultGroups) {
-  const uint16_t kDefaultGroups[] = {SSL_GROUP_X25519_MLKEM768,
-                                     SSL_GROUP_X25519, SSL_GROUP_SECP256R1,
-                                     SSL_GROUP_SECP384R1};
+  const uint16_t kDefaultGroups[] = {
+      SSL_GROUP_X25519_MLKEM768, SSL_GROUP_P256_KYBER768_DRAFT00,
+      SSL_GROUP_X25519, SSL_GROUP_SECP256R1, SSL_GROUP_SECP384R1};
   const uint32_t kBogusFlags[] = {SSL_GROUP_FLAG_EQUAL_PREFERENCE_WITH_NEXT,
                                   SSL_GROUP_FLAG_EQUAL_PREFERENCE_WITH_NEXT, 0};
   bssl::UniquePtr<SSL_CTX> server_ctx =
diff --git a/include/openssl/prefix_symbols.h b/include/openssl/prefix_symbols.h
index 79df39104..66d415d93 100644
--- a/include/openssl/prefix_symbols.h
+++ b/include/openssl/prefix_symbols.h
@@ -2099,6 +2099,7 @@
 #pragma redefine_extname SSL_CTX_use_certificate_chain_file BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_CTX_use_certificate_chain_file)
 #pragma redefine_extname SSL_CTX_use_certificate_file BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_CTX_use_certificate_file)
 #pragma redefine_extname SSL_CTX_use_psk_identity_hint BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_CTX_use_psk_identity_hint)
+#pragma redefine_extname SSL_CTX_use_second_keyshare BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_CTX_use_second_keyshare)
 #pragma redefine_extname SSL_ECH_KEYS_add BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_ECH_KEYS_add)
 #pragma redefine_extname SSL_ECH_KEYS_free BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_ECH_KEYS_free)
 #pragma redefine_extname SSL_ECH_KEYS_has_duplicate_config_id BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_ECH_KEYS_has_duplicate_config_id)
@@ -2446,6 +2447,7 @@
 #pragma redefine_extname SSL_use_certificate_ASN1 BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_use_certificate_ASN1)
 #pragma redefine_extname SSL_use_certificate_file BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_use_certificate_file)
 #pragma redefine_extname SSL_use_psk_identity_hint BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_use_psk_identity_hint)
+#pragma redefine_extname SSL_use_second_keyshare BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_use_second_keyshare)
 #pragma redefine_extname SSL_used_hello_retry_request BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_used_hello_retry_request)
 #pragma redefine_extname SSL_version BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_version)
 #pragma redefine_extname SSL_want BORINGSSL_ADD_USER_LABEL_AND_PREFIX(SSL_want)
@@ -5233,6 +5235,7 @@
 #define SSL_CTX_use_certificate_chain_file BORINGSSL_ADD_PREFIX(SSL_CTX_use_certificate_chain_file)
 #define SSL_CTX_use_certificate_file BORINGSSL_ADD_PREFIX(SSL_CTX_use_certificate_file)
 #define SSL_CTX_use_psk_identity_hint BORINGSSL_ADD_PREFIX(SSL_CTX_use_psk_identity_hint)
+#define SSL_CTX_use_second_keyshare BORINGSSL_ADD_PREFIX(SSL_CTX_use_second_keyshare)
 #define SSL_ECH_KEYS_add BORINGSSL_ADD_PREFIX(SSL_ECH_KEYS_add)
 #define SSL_ECH_KEYS_free BORINGSSL_ADD_PREFIX(SSL_ECH_KEYS_free)
 #define SSL_ECH_KEYS_has_duplicate_config_id BORINGSSL_ADD_PREFIX(SSL_ECH_KEYS_has_duplicate_config_id)
@@ -5580,6 +5583,7 @@
 #define SSL_use_certificate_ASN1 BORINGSSL_ADD_PREFIX(SSL_use_certificate_ASN1)
 #define SSL_use_certificate_file BORINGSSL_ADD_PREFIX(SSL_use_certificate_file)
 #define SSL_use_psk_identity_hint BORINGSSL_ADD_PREFIX(SSL_use_psk_identity_hint)
+#define SSL_use_second_keyshare BORINGSSL_ADD_PREFIX(SSL_use_second_keyshare)
 #define SSL_used_hello_retry_request BORINGSSL_ADD_PREFIX(SSL_used_hello_retry_request)
 #define SSL_version BORINGSSL_ADD_PREFIX(SSL_version)
 #define SSL_want BORINGSSL_ADD_PREFIX(SSL_want)
